# IDS Assignment Part 2 - <font color="red"><h7>Deadline: 23/01/2023 23:59</h7></font>
This is the second part of the assignment in IDS 2022/2023. 
Please use this Jupyter notebook to work on the questions posed in the assignment. When you are done, upload the notebook in Moodle at the designated activity. In addition to the _Jupyter notebook_, please submit _one zip-file_ containing your screenshots for Question 7. 

Give your commented Python code and answers in the corresponding provided cells. Make sure to answer all questions in a clear and explicit manner and discuss your outputs. _Please do not change the general structure of this notebook_. You can, however, add additional markdown or code cells if necessary. <b>Please DO NOT CLEAR THE OUTPUT of the notebook you are submitting! </b>

<font color="red"> *Please make sure to include the names and matriculation numbers of all group members in the slot provided below.* </font> If a name or a student id is missing, the student will not receive any points.

Hint 1: While working on the assignment, you will get a better understanding of the dataset. Feel free to generate additional results and visualizations to support your answers. For example, this might be useful regarding data modification, data simplification, or output interpretation. <font color="red">Ensure that all your claims are supported.</font>

Hint 2: <font color="red">Plan your time wisely. </font> A few parts of this assignment may take some time to run. It might be necessary to consider time management when you plan your group work. Also, do not attempt to upload your assignment at the last minute before the deadline. This often does not work, and you will miss the deadline. Late submissions will not be considered.

Hint 3: RWTHmoodle allows multiple submissions, with every new submission overwriting the previous one. <b>Partial submissions are therefore possible and encouraged. </b> This might be helpful in case of technical issues with RWTHMoodle, which may occur close to the deadline.

<font color="red"><b>Student Names and IDs:
    
    1. Minh Nghia Phan 394806
    
    2. 
    
    3.

## Question 1: Preprocessing (20 points)

In [165]:
import pandas as pd
import numpy as np

### Sklearn
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.impute import KNNImputer

In this question, we consider a dataset documenting the Ski Resorts in Europe (**ski_resorts.csv**).
Each row contains some information about the Ski resort.
You can find a short description for each column:

| Column | Description |
| ------ | ----------- |
| Resort | The name of the ski & snowboard resort. |
| Country | The country in which the resort is located. |
| HighestPoint | The highest mountain point at the ski resort.   |
| LowestPoint | The lowest possible point to ski at the ski resort.  |
| DayPassPriceAdult | The price shows what it costs for 1 adult for 1 day in the main season in Euro. |
| BeginnerSlope | The total amount of “beginner” slopes in kilometer at the resort. “Beginner slopes” contains “children”, “blue” and, “green” slopes. |
| IntermediateSlope | The total amount of “intermediate” slopes in kilometer at the resort. “Intermediate slopes” contains “red” slopes. |
| DifficultSlope | The total amount of “difficult” slopes in kilometer at the resort. “Difficult slopes” contains “black”, “advanced”, and ”expert” slopes. |
| TotalSlope | The sum of “beginner slopes” + “intermediate slopes” + “difficult slopes” |
| Snowparks | Does the resort have one or more snowparks, or not? |
| NightSki | Does the resort offer skiing on illuminated slopes? |
| SurfaceLifts | The amount of lifts in this category: T-bar, Sunkidslift, Rope lifts, and people mower. |
| ChairLifts | The total amount of chairlifts. |
| GondolaLifts | The amount of lifts in this category: Gondola, Train lifts, Funicular, Combined gondola and chairlifts, Helicopter lifts, Snowcats, and Aerial tramways. |
| TotalLifts | The sum of “surface lifts etc” + “gondola etc” + “chairlifts etc.” |
| LiftCapacity | How many passengers can the lift system at the ski resort mowe in one hour? |
| SnowCannons  |The total amount of snow cannons at the ski resort.|

### a) Loading the Data and Initial Quality Investigation (2.5pts)

#### **a(i)** 
Load the dataset into a dataframe `df`. <font color='red'>Use the first column as index for your dataframe</font>. Ensure that the index is valid, that is, it should not contain any duplicate entries. 



**In the subsequent questions, only modify the dataframe `df` if explicitly requested. However, you can always create working copies.**

In [166]:
df = pd.read_csv('datasets/ski_resorts.csv', index_col=0)
df

Resort          Country  \
1                                Alpendorf (Ski amedé)          Austria   
2    Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...          Andorra   
3                                Oberau (Wildschönau)          Austria   
4                                       Dachstein West          Austria   
5                                          Rosa Khutor  Southern Russia   
..                                                 ...              ...   
372                          Montgenèvre (Via Lattea)           France   
373                          Sauze d’Oulx (Via Lattea)            Italy   
374             Gressoney - La-Trinite (Monterosa Ski)            Italy   
375                          Champoluc (Monterosa Ski)            Italy   
376                                         Zauchensee          Austria   

     HighestPoint  LowestPoint  DayPassPriceAdult  BeginnerSlope  \
1          1980.0        740.0               52.0           30.0   
2          2640.0       1710.0               47.0          100.0   
3          1130.0        900.0               30.0            1.0   
4          1620.0        780.0               42.0           15.0   
5          2320.0        940.0               22.0           30.0   
..            ...          ...                ...            ...   
372        2749.0       1372.0               48.0           96.0   
373        2749.0       1372.0               48.0           96.0   
374        3275.0       1212.0               43.0           23.0   
375        3275.0       1212.0               43.0           23.0   
376        2188.0       1000.0               52.0           23.0   

     IntermediateSlope  DifficultSlope  TotalSlope Snowparks NightSki  \
1                   81             4.0         115       Yes       No   
2                   77            33.0         210       Yes      Yes   
3                    0             1.0           2        No       No   
4                   33             3.0          51       Yes      Yes   
5                   26            21.0          77       Yes       No   
..                 ...             ...         ...       ...      ...   
372                220            84.0         400        No      Yes   
373                220            84.0         400        No      Yes   
374                 94            15.0         132       Yes       No   
375                 94            15.0         132       Yes       No   
376                 16             4.0          44       Yes       No   

     SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  LiftCapacity  \
1            22.0          16          11.0          49       75398.0   
2            37.0          28           7.0          72       99017.0   
3             2.0           0           0.0           2        1932.0   
4            25.0           8           3.0          36       32938.0   
5             6.0          11          10.0          27       49228.0   
..            ...         ...           ...         ...           ...   
372           NaN          35           7.0          71       96433.0   
373          29.0          35           7.0          71       96433.0   
374           9.0           9          12.0          30       31984.0   
375           9.0           9          12.0          30       31984.0   
376           9.0           6           4.0          19       25988.0   

     SnowCannons  
1          600.0  
2         1032.0  
3            0.0  
4          163.0  
5          450.0  
..           ...  
372          0.0  
373          0.0  
374        655.0  
375        655.0  
376        113.0  

[376 rows x 17 columns]

#### **a(ii)** 
Show the data types of the dataframe columns as well as the first 5 rows. On the first sight, are there any data type problems (e.g., numerical columns having a non-numerical data type)?



In [167]:
# Show column data types
df.dtypes

Resort                object
Country               object
HighestPoint         float64
LowestPoint          float64
DayPassPriceAdult    float64
BeginnerSlope        float64
IntermediateSlope      int64
DifficultSlope       float64
TotalSlope             int64
Snowparks             object
NightSki              object
SurfaceLifts         float64
ChairLifts             int64
GondolaLifts         float64
TotalLifts             int64
LiftCapacity         float64
SnowCannons          float64
dtype: object

In [168]:
# First 5 rows
df.head()

Resort          Country  \
1                              Alpendorf (Ski amedé)          Austria   
2  Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...          Andorra   
3                              Oberau (Wildschönau)          Austria   
4                                     Dachstein West          Austria   
5                                        Rosa Khutor  Southern Russia   

   HighestPoint  LowestPoint  DayPassPriceAdult  BeginnerSlope  \
1        1980.0        740.0               52.0           30.0   
2        2640.0       1710.0               47.0          100.0   
3        1130.0        900.0               30.0            1.0   
4        1620.0        780.0               42.0           15.0   
5        2320.0        940.0               22.0           30.0   

   IntermediateSlope  DifficultSlope  TotalSlope Snowparks NightSki  \
1                 81             4.0         115       Yes       No   
2                 77            33.0         210       Yes      Yes   
3                  0             1.0           2        No       No   
4                 33             3.0          51       Yes      Yes   
5                 26            21.0          77       Yes       No   

   SurfaceLifts  ChairLifts  GondolaLifts  TotalLifts  LiftCapacity  \
1          22.0          16          11.0          49       75398.0   
2          37.0          28           7.0          72       99017.0   
3           2.0           0           0.0           2        1932.0   
4          25.0           8           3.0          36       32938.0   
5           6.0          11          10.0          27       49228.0   

   SnowCannons  
1        600.0  
2       1032.0  
3          0.0  
4        163.0  
5        450.0

**Your Answer:** 
1. The columns **IntermediateSlope** and **TotalSlope** measure the amount of slopes in kilometers and therefore should have types __float64__ instead of __int64__.

2. **Snowparks** and **NightSki** should have __boolean__ type because they only have *Yes* and *No* values.

3. **SurfaceLifts**, **GondolaLifts**, **LiftCapacity**, and **SnowCannons** should have type __int64__ instead of __float64__ because they indicate either number of lifts or number of passengers, which should all be integers.

#### **a(iii)** 
To improve performance and memory usage (in particular for large datasets) it is important to use **categorical** columns whenever suitable.
Are there any categorical column candidates? Explain your answer. \
Afterward, convert the column(s) in `df` into categorical column(s).

In [169]:
# Convert to categorical
tocat = ['Country', 'Snowparks', 'NightSki']
for col in tocat:
    df[col] = df[col].astype('category')

In [170]:
# Number of unique resorts
len(df['Resort'].unique())

376

**Your Answer:** **Country** can be categorical, as each country is a category itself.

The other columns, other than **Resort**, **Snowparks** and **NightSki**, can not be categorical, because they represent numerical real values and measure magnitudes (kilometers, prices, number of lifts/passengers).

**Snowparks** and **NightSki** should be boolean instead of categorical, because they have only 2 values "Yes" and "No".

**Resort** represents the unique name of each resort in each row (there are 376 unique resorts corresponding to the exact 376 rows of the dataframe), hence it should not be categorical.

### b) Handling Missing Values & Encoding (17.5pts)

In the following task, you can assume that every NAN entry in the dataframe is actually a missing value. This can partially be justified by the fact that pandas did not have problems inferring the "proper" datatypes (e.g., a string indicating a missing number in a number column would result in pandas parsing an object column) and your subsequent check of the data types. Therefore, you can use `df.isna()` as a proxy indicator for missing values.

#### **b(i)** 
Simply discarding missing entries is usually not a good idea. Therefore, you should first analyze the number of missing values and check for patterns of missing values. 

To this end, compute the following statistics on missing values:
1. How many entries does the dataframe have? (To relate this to the number of entries missing)
2. How many missing values do we have? What is the ratio i.e., "number of missing values"/"number of entries of df"?
3. How many rows have at least a single missing value?
4. Count the number of missing values per column.
5. Count the number of missing values per row and aggregate them - i.e., show the number of rows that suffer from x missing values.
6. What do you observe? Are there any rows containing missing values for the same set of columns? Can you identify potential patterns?


In [171]:
# your code for 1. How many entries does the dataframe have? (To relate this to the number of entries missing)
n_entries = len(df.columns)*len(df.index)
print(f"Number of entries in the dataframe: {n_entries}")

Number of entries in the dataframe: 6392


In [172]:
# your code for 2. How many missing values do we have? What is the ratio i.e., "number of missing values"/"number of entries of df"?
n_missing = df.isnull().sum().sum()
print(f"Number of missing values in the dataframe: {n_missing}")
print(f"Ratio of missing values: {n_missing/n_entries}")

Number of missing values in the dataframe: 87
Ratio of missing values: 0.013610763454317897


In [173]:
# your code for 3. How many rows have at least a single missing value?
df_rows_na = df[df.isna().any(axis=1)] # DF of rows with at least 1 NaN value
print(f"Number of rows with at least 1 missing value: {len(df_rows_na)}")

Number of rows with at least 1 missing value: 45


In [174]:
# your code for 4. Count the number of missing values per column.
print(f"Number of missing values per columns")
df.isnull().sum()

Number of missing values per columns


Resort                0
Country               0
HighestPoint         13
LowestPoint           3
DayPassPriceAdult    17
BeginnerSlope         8
IntermediateSlope     0
DifficultSlope        8
TotalSlope            0
Snowparks             0
NightSki              0
SurfaceLifts          6
ChairLifts            0
GondolaLifts          7
TotalLifts            0
LiftCapacity         17
SnowCannons           8
dtype: int64

In [175]:
# your code for 5. Count the number of missing values per row and aggregate them - i.e., show the number of rows that suffer from x missing values.
df_missing = df.isna().astype("int") # entry=1 indicates missing values, else 0
df_missing = df_missing.apply(np.sum, axis=1) # sum all rows
df_missing = df_missing.value_counts() # aggregate by num of missing values
print('Left column: number of missing values, right: number of rows with that many missing values')
df_missing

Left column: number of missing values, right: number of rows with that many missing values


0    331
1     35
6      8
2      2
dtype: int64

**Your answer:** *(for 6. What do you observe? Are there any rows containing missing values for the same set of columns?)*

We observe that there are 8 rows with 6 missing values. From 4., we can see that there are exactly 6 columns with at least 8 missing values (HighestPoint, DayPassPriceAdult, BeginnerSlope, DifficultSlope, LiftCapacity, and SnowCannons). So the missing values of these rows are all from these 6 columns.

#### **b(ii)**
For the next step:

1. Transform the categorical column(s) you identified in a(iii) into one-hot encoding format. 
2. Transform the columns "Snowparks" and "NightSki" in `df` into boolean data type, where "Yes" should be `True` and "No" should be `False`

In the end, the original categorical column(s) should still be there. Additionally, there should be x number (x is the number of unique values) of one-hot encoding columns for each categorical column. Use the following naming convention for the new columns "{name of the categorical column}_{unique value for that column}" Also, make sure the columns "Snowparks" and "NightSki" are boolean type in the end.
Lastly, print the top five rows of the resulting dataframe.

*Hint: You can use the pd.get_dummies() function from pandas for the first transformation*

In [176]:
# your code for 1. Transform the categorical column(s) you identified in a(iii) into one-hot encoding format. 
df2 = pd.get_dummies(df, columns=['Country'])
df2['Country'] = df['Country']
df2

Resort  HighestPoint  \
1                                Alpendorf (Ski amedé)        1980.0   
2    Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...        2640.0   
3                                Oberau (Wildschönau)        1130.0   
4                                       Dachstein West        1620.0   
5                                          Rosa Khutor        2320.0   
..                                                 ...           ...   
372                          Montgenèvre (Via Lattea)        2749.0   
373                          Sauze d’Oulx (Via Lattea)        2749.0   
374             Gressoney - La-Trinite (Monterosa Ski)        3275.0   
375                          Champoluc (Monterosa Ski)        3275.0   
376                                         Zauchensee        2188.0   

     LowestPoint  DayPassPriceAdult  BeginnerSlope  IntermediateSlope  \
1          740.0               52.0           30.0                 81   
2         1710.0               47.0          100.0                 77   
3          900.0               30.0            1.0                  0   
4          780.0               42.0           15.0                 33   
5          940.0               22.0           30.0                 26   
..           ...                ...            ...                ...   
372       1372.0               48.0           96.0                220   
373       1372.0               48.0           96.0                220   
374       1212.0               43.0           23.0                 94   
375       1212.0               43.0           23.0                 94   
376       1000.0               52.0           23.0                 16   

     DifficultSlope  TotalSlope Snowparks NightSki  ...  Country_Siberia  \
1               4.0         115       Yes       No  ...                0   
2              33.0         210       Yes      Yes  ...                0   
3               1.0           2        No       No  ...                0   
4               3.0          51       Yes      Yes  ...                0   
5              21.0          77       Yes       No  ...                0   
..              ...         ...       ...      ...  ...              ...   
372            84.0         400        No      Yes  ...                0   
373            84.0         400        No      Yes  ...                0   
374            15.0         132       Yes       No  ...                0   
375            15.0         132       Yes       No  ...                0   
376             4.0          44       Yes       No  ...                0   

     Country_Slovakia  Country_Slovenia  Country_Southern Russia  \
1                   0                 0                        0   
2                   0                 0                        0   
3                   0                 0                        0   
4                   0                 0                        0   
5                   0                 0                        1   
..                ...               ...                      ...   
372                 0                 0                        0   
373                 0                 0                        0   
374                 0                 0                        0   
375                 0                 0                        0   
376                 0                 0                        0   

     Country_Spain  Country_Sweden  Country_Switzerland  Country_Ukraine  \
1                0               0                    0                0   
2                0               0                    0                0   
3                0               0                    0                0   
4                0               0                    0                0   
5                0               0                    0                0   
..             ...             ...                  ...              ...   
372              0               0          

In [177]:
# your code for 2. Transform the columns "Snowparks" and "NightSki" in `df` into boolean data type, where "Yes" should be `True` and "No" should be `False`
to_bool = ['Snowparks', 'NightSki']
for col in to_bool:
    df2[col] = (df2[col] == 'Yes')
df2

Resort  HighestPoint  \
1                                Alpendorf (Ski amedé)        1980.0   
2    Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...        2640.0   
3                                Oberau (Wildschönau)        1130.0   
4                                       Dachstein West        1620.0   
5                                          Rosa Khutor        2320.0   
..                                                 ...           ...   
372                          Montgenèvre (Via Lattea)        2749.0   
373                          Sauze d’Oulx (Via Lattea)        2749.0   
374             Gressoney - La-Trinite (Monterosa Ski)        3275.0   
375                          Champoluc (Monterosa Ski)        3275.0   
376                                         Zauchensee        2188.0   

     LowestPoint  DayPassPriceAdult  BeginnerSlope  IntermediateSlope  \
1          740.0               52.0           30.0                 81   
2         1710.0               47.0          100.0                 77   
3          900.0               30.0            1.0                  0   
4          780.0               42.0           15.0                 33   
5          940.0               22.0           30.0                 26   
..           ...                ...            ...                ...   
372       1372.0               48.0           96.0                220   
373       1372.0               48.0           96.0                220   
374       1212.0               43.0           23.0                 94   
375       1212.0               43.0           23.0                 94   
376       1000.0               52.0           23.0                 16   

     DifficultSlope  TotalSlope  Snowparks  NightSki  ...  Country_Siberia  \
1               4.0         115       True     False  ...                0   
2              33.0         210       True      True  ...                0   
3               1.0           2      False     False  ...                0   
4               3.0          51       True      True  ...                0   
5              21.0          77       True     False  ...                0   
..              ...         ...        ...       ...  ...              ...   
372            84.0         400      False      True  ...                0   
373            84.0         400      False      True  ...                0   
374            15.0         132       True     False  ...                0   
375            15.0         132       True     False  ...                0   
376             4.0          44       True     False  ...                0   

     Country_Slovakia  Country_Slovenia  Country_Southern Russia  \
1                   0                 0                        0   
2                   0                 0                        0   
3                   0                 0                        0   
4                   0                 0                        0   
5                   0                 0                        1   
..                ...               ...                      ...   
372                 0                 0                        0   
373                 0                 0                        0   
374                 0                 0                        0   
375                 0                 0                        0   
376                 0                 0                        0   

     Country_Spain  Country_Sweden  Country_Switzerland  Country_Ukraine  \
1                0               0                    0                0   
2                0               0                    0                0   
3                0               0                    0                0   
4                0               0                    0                0   
5                0               0                    0                0   
..             ...             ...                  ...              ...   
372              0  

In [178]:
# use df.dtypes to check if you correctly transform the data
df2.dtypes

Resort                              object
HighestPoint                       float64
LowestPoint                        float64
DayPassPriceAdult                  float64
BeginnerSlope                      float64
IntermediateSlope                    int64
DifficultSlope                     float64
TotalSlope                           int64
Snowparks                             bool
NightSki                              bool
SurfaceLifts                       float64
ChairLifts                           int64
GondolaLifts                       float64
TotalLifts                           int64
LiftCapacity                       float64
SnowCannons                        float64
Country_Andorra                      uint8
Country_Austria                      uint8
Country_Bosnia and Herzegovina       uint8
Country_Bulgaria                     uint8
Country_Czech Republic               uint8
Country_Denmark                      uint8
Country_Finland                      uint8
Country_Fra

In [179]:
# First 5 rows
df2.head()

Resort  HighestPoint  \
1                              Alpendorf (Ski amedé)        1980.0   
2  Soldeu-Pas de la Casa/​Grau Roig/​El Tarter/​C...        2640.0   
3                              Oberau (Wildschönau)        1130.0   
4                                     Dachstein West        1620.0   
5                                        Rosa Khutor        2320.0   

   LowestPoint  DayPassPriceAdult  BeginnerSlope  IntermediateSlope  \
1        740.0               52.0           30.0                 81   
2       1710.0               47.0          100.0                 77   
3        900.0               30.0            1.0                  0   
4        780.0               42.0           15.0                 33   
5        940.0               22.0           30.0                 26   

   DifficultSlope  TotalSlope  Snowparks  NightSki  ...  Country_Siberia  \
1             4.0         115       True     False  ...                0   
2            33.0         210       True      True  ...                0   
3             1.0           2      False     False  ...                0   
4             3.0          51       True      True  ...                0   
5            21.0          77       True     False  ...                0   

   Country_Slovakia  Country_Slovenia  Country_Southern Russia  Country_Spain  \
1                 0                 0                        0              0   
2                 0                 0                        0              0   
3                 0                 0                        0              0   
4                 0                 0                        0              0   
5                 0                 0                        1              0   

   Country_Sweden  Country_Switzerland  Country_Ukraine  \
1               0                    0                0   
2               0                    0                0   
3               0                    0                0   
4               0                    0                0   
5               0                    0                0   

   Country_United Kingdom          Country  
1                       0          Austria  
2                       0          Andorra  
3                       0          Austria  
4                       0          Austria  
5                       0  Southern Russia  

[5 rows x 44 columns]

#### **b(iii)** 
The previous analysis in b(i) showed that there are missing values in the 'SurfaceLifts' and 'GondolaLifts' columns.\
How would you impute these values? \
Motivate your approach and apply it to `df`.

*Hint: Remember the semantics of the columns. Also, carefully assert your assumptions.*


**Your Answer:** *(Motivate your approach.)* From the description, TotalLifts is the sum of SurfaceLifts + GondolaLifts + ChairLifts. Luckily, the code below shows that there is no row where both the values of GondolaLifts and SurfaceLifts are missing. Also from exercise **b(i)**, there is no missing value in ChairLifts. So we can impute these as follows:
- Where GondolaLifts is missing, impute it with TotalLifts - SurfaceLifts - ChairLifts.
- Where SurfaceLifts is missing, impute it with TotalLifts - GondolaLifts - ChairLifts.

In [180]:
# your code
gondola = df2[df2['GondolaLifts'].isna()].copy()
gondola['GondolaLifts'] = gondola['TotalLifts'] - gondola['ChairLifts'] - gondola['SurfaceLifts']
surface = df2[df2['SurfaceLifts'].isna()].copy()
surface['SurfaceLifts'] = surface['TotalLifts'] - surface['ChairLifts'] - surface['GondolaLifts']
df2.loc[gondola.index, ['GondolaLifts']] = gondola['GondolaLifts']
df2.loc[surface.index, ['SurfaceLifts']] = surface['SurfaceLifts']

#### **b(iv)**
Impute the rest of the missing values using the knn-imputation method. To this end, apply the following steps:
1. Create a working copy `df_tmp` of your updated `df`.
2. For simplicity, drop the non-numerical columns (i.e., not of types integer or floats), which also inlcude the one-hot encoded and the boolean columns* you created earlier.
3. Normalize the data in `df_tmp` (e.g., Standard score normalization). If the features have very different scales, knn can become very biased.
4. Impute the missing values considering six neighbors.
5. Invert the transformation applied upfront to enable more meaningful and intuitive visualizations.
6. Append the columns you dropped in step 2.
 
In the end, `df` should not contain missing values and have all the columns.

\*Note that by dropping the columns we lose the information of countries and the two boolean attributes ("Snowparks" and "NightSki") when imputing the missing values, which might be crucial for inferencing values such as the price for a ski pass. In practice, one should try to find if there are correlations before deciding whether to drop the columns or not.
We drop the columns here to make the following steps easier because we only have to deal with numerical columns.

*Hint: Be careful with the indices of your dataframes.*


In [181]:
# your code for step 1 & 2
df_tmp = df2.copy()
df_tmp = df_tmp[df.columns]
dropped_cols = ['Country', 'Resort', 'NightSki', 'Snowparks']
df_tmp = df_tmp.drop(dropped_cols, axis=1)
df_tmp.columns

Index(['HighestPoint', 'LowestPoint', 'DayPassPriceAdult', 'BeginnerSlope',
       'IntermediateSlope', 'DifficultSlope', 'TotalSlope', 'SurfaceLifts',
       'ChairLifts', 'GondolaLifts', 'TotalLifts', 'LiftCapacity',
       'SnowCannons'],
      dtype='object')

In [182]:
# your code for step 3
scaler = MinMaxScaler()
df_scaled = scaler.fit_transform(df_tmp)
df_scaled

array([[0.49192481, 0.33944954, 0.64197531, ..., 0.2816092 , 0.29886634,
        0.25178347],
       [0.66666667, 0.78440367, 0.58024691, ..., 0.4137931 , 0.3924885 ,
        0.43306756],
       [0.26687847, 0.41284404, 0.37037037, ..., 0.01149425, 0.00765816,
        0.        ],
       ...,
       [0.83478952, 0.5559633 , 0.5308642 , ..., 0.17241379, 0.12677977,
        0.27486362],
       [0.83478952, 0.5559633 , 0.5308642 , ..., 0.17241379, 0.12677977,
        0.27486362],
       [0.54699497, 0.4587156 , 0.64197531, ..., 0.1091954 , 0.10301253,
        0.04741922]])

In [183]:
# your code for step 4
imputer = KNNImputer(n_neighbors=6)
df_scaled_imputed = imputer.fit_transform(df_scaled)
df_scaled_imputed

array([[0.49192481, 0.33944954, 0.64197531, ..., 0.2816092 , 0.29886634,
        0.25178347],
       [0.66666667, 0.78440367, 0.58024691, ..., 0.4137931 , 0.3924885 ,
        0.43306756],
       [0.26687847, 0.41284404, 0.37037037, ..., 0.01149425, 0.00765816,
        0.        ],
       ...,
       [0.83478952, 0.5559633 , 0.5308642 , ..., 0.17241379, 0.12677977,
        0.27486362],
       [0.83478952, 0.5559633 , 0.5308642 , ..., 0.17241379, 0.12677977,
        0.27486362],
       [0.54699497, 0.4587156 , 0.64197531, ..., 0.1091954 , 0.10301253,
        0.04741922]])

In [184]:
# your code for step 5
df_scaled_imputed_inverted = scaler.inverse_transform(df_scaled_imputed)
df_scaled_imputed_inverted

array([[1.9800e+03, 7.4000e+02, 5.2000e+01, ..., 4.9000e+01, 7.5398e+04,
        6.0000e+02],
       [2.6400e+03, 1.7100e+03, 4.7000e+01, ..., 7.2000e+01, 9.9017e+04,
        1.0320e+03],
       [1.1300e+03, 9.0000e+02, 3.0000e+01, ..., 2.0000e+00, 1.9320e+03,
        0.0000e+00],
       ...,
       [3.2750e+03, 1.2120e+03, 4.3000e+01, ..., 3.0000e+01, 3.1984e+04,
        6.5500e+02],
       [3.2750e+03, 1.2120e+03, 4.3000e+01, ..., 3.0000e+01, 3.1984e+04,
        6.5500e+02],
       [2.1880e+03, 1.0000e+03, 5.2000e+01, ..., 1.9000e+01, 2.5988e+04,
        1.1300e+02]])

In [190]:
# your code for step 6
# To make rerunning cells not a pain, we store the final result in a new Dataframe dff
dff = df_tmp.copy()
dff.loc[:,scaler.feature_names_in_] = df_scaled_imputed_inverted
dff = pd.concat([dff, df2[df2.filter(regex=("Country_*")).columns]], axis=1) # Reappend the one-hot columns
dff[dropped_cols] = df2[dropped_cols] # Reappend the dropped categorical columns 
dff # Final result

HighestPoint  LowestPoint  DayPassPriceAdult  BeginnerSlope  \
1          1980.0        740.0               52.0           30.0   
2          2640.0       1710.0               47.0          100.0   
3          1130.0        900.0               30.0            1.0   
4          1620.0        780.0               42.0           15.0   
5          2320.0        940.0               22.0           30.0   
..            ...          ...                ...            ...   
372        2749.0       1372.0               48.0           96.0   
373        2749.0       1372.0               48.0           96.0   
374        3275.0       1212.0               43.0           23.0   
375        3275.0       1212.0               43.0           23.0   
376        2188.0       1000.0               52.0           23.0   

     IntermediateSlope  DifficultSlope  TotalSlope  SurfaceLifts  ChairLifts  \
1                 81.0             4.0       115.0          22.0        16.0   
2                 77.0            33.0       210.0          37.0        28.0   
3                  0.0             1.0         2.0           2.0         0.0   
4                 33.0             3.0        51.0          25.0         8.0   
5                 26.0            21.0        77.0           6.0        11.0   
..                 ...             ...         ...           ...         ...   
372              220.0            84.0       400.0          29.0        35.0   
373              220.0            84.0       400.0          29.0        35.0   
374               94.0            15.0       132.0           9.0         9.0   
375               94.0            15.0       132.0           9.0         9.0   
376               16.0             4.0        44.0           9.0         6.0   

     GondolaLifts  ...  Country_Southern Russia  Country_Spain  \
1            11.0  ...                        0              0   
2             7.0  ...                        0              0   
3             0.0  ...                        0              0   
4             3.0  ...                        0              0   
5            10.0  ...                        1              0   
..            ...  ...                      ...            ...   
372           7.0  ...                        0              0   
373           7.0  ...                        0              0   
374          12.0  ...                        0              0   
375          12.0  ...                        0              0   
376           4.0  ...                        0              0   

     Country_Sweden  Country_Switzerland  Country_Ukraine  \
1                 0                    0                0   
2                 0                    0                0   
3                 0                    0                0   
4                 0                    0                0   
5                 0                    0                0   
..              ...                  ...              ...   
372               0                    0                0   
373               0                    0                0   
374               0                    0                0   
375               0                    0                0   
376               0                    0                0   

     Country_United Kingdom          Country  \
1                         0          Austria   
2                         0          Andorra   
3                         0          Austria   
4                         0          Austria   
5                         0  Southern Russia   
..                      ...              ...   
372                       0           France   
373                       0            Italy   
374                       0            Italy   
375                       0            Italy   
376                       0          Austria   

                                                Resort  NightSki  Snowparks  
1                                Alpendorf (Sk

In [193]:
# assert df.isna().sum().sum() == 0
assert dff.isna().sum().sum() == 0
dff.isna().sum().sum() # To believe it, we have to see it

0

## Question 2: Visualization (13 points)

In this task, you will analyze the data using different means of visualization.

Start with the following preprocessed and integrated dataframe `df_v`. \
Note that it has a similar structure to the dataframe that you should obtain from the previous task, however, the values have been modified.

In [ ]:
import pandas as pd
import plotly.express as px
from pandas.plotting import scatter_matrix
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df_v = pd.read_csv("./datasets/ski_resorts_visual.csv", index_col=0)

### **a) (3pts)** 
To start the visual analysis, make a Scatter plot matrix to visually check if there are any correlations between the numerical attributes.

*Hint: You can use the scatter_matrix from pandas.plotting or pairplot from seaborn to make the plot.*


In [ ]:
# your code

### **b) (3pts)**
Another way to detect correlation is to calculate the Pearson correlation coefficient. Calculate the correlation matrix for the numerical data and visualize the matrix using a heatmap. 
Briefly discuss your findings from the heatmap and the scatter plot you created in 2(a).

Make sure to annotate the heatmap with the values of the correlation.

*Hint: You can use the heatmap function from seaborn to make the plot.*

In [ ]:
# your code

### **c) (4pts)** 
We now want to have an overview of the attribute "TotalSlope" aggregated by different levels of hierarchy (Europe -> Country -> Resort). It seems that a tree map is suitable for this purpose.

Make a tree map where
- the root node represents Europe.
- the child nodes of Europe are countries.
- the child nodes of each country are the ski resorts.
- the size of the rectangles is determined by the attribute "TotalSlope".

Also, use the tree map to find out
1. The sum of TotalSlopes of a country, list the top five countries and the corresponding values.
2. The max value of TotalSlope of the five countries you identified in 1.

*Hint: You can use the treemap function from plotly.express.*

In [ ]:
# your code

**Your answer for...** \
*...  1. The sum of TotalSlopes of a country. List the top five countries and the corresponding values:* \
*...  2. The max value of TotalSlope of the five countries you identified in 1:*


### **d) (3pts)** 

An alternative to a tree map is a sunburst plot, the principle is similar to a tree map. 

Recall from the lecture that:
- Each ring is a different level of the hierarchy
- Each segment of a ring belongs to one categorical value
- The size of a segment is either divided proportionally to a value

Now, we would like to have an overview of the attribute "TotalLifts" aggregated by different level of hierarchy.

Make a sunburst plot where
- the first hierarchy(ring) is "Country"
- the second hierarchy(ring) is "Snowparks" (whether the resort has snowparks)
- the third hierarchy(ring) is "Resort"
- the size of the segments is determined by the attribute "TotalLifts".

Then, briefly discuss your findings from the plot.

*Hint: You can use the sunburst function from plotly.express.*

In [ ]:
# your code

**Your answer:**

## Question 3 - Frequent Item Sets and Association Rules (15 points)

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules as arule

### A real online retail transaction data set of two years.

Data Set Information:
This Retail dataset contains all the transactions occurring for a UK-based and registered, non-store online retail between 01/12/2009 and 09/12/2011.The company mainly sells unique all-occasion gift ware. Many customers of the company are wholesalers.

Attribute Information:
- Invoice: Invoice number. Nominal. A 6-digit integral number is uniquely assigned to each transaction. If the number starts with 'C' it refers to a canceled transaction.
- StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.
- Description: Product (item) name. Nominal.
- Quantity: The quantities of each product (item) per transaction. Numeric.
- InvoiceDate: Invoice date and time. Numeric. The day and time when a transaction was generated.
- Price: Unit price. Numeric. Product price per unit in sterling (£).
- CustomerID: Customer number. Nominal. A 5-digit integral number is uniquely assigned to each customer. This number has postfix 'n'.
- Country: Country name. Nominal. The name of the country where a customer resides.

### a) Loading, exploring and preprocessing the dataset (4 pts)

#### **a(i)** 
Load the data from `retail.csv` and save it under the variable `retail_df`. Display the first few lines of the dataframe.

In [ ]:
# your code

#### **a(ii)** 
To get to know the dataset, do the following:

- Show the number of rows in the dataset.
- Show the number of unique customers.
- Show the number of unique product names.
- Show the number of unique invoices.
- Show the number and the list of all the countries where the customers reside.

In [ ]:
# your code

#### **a(iii)** 
You are interested in analyzing itemsets that are frequently purchased together. Before continuing with that task, you have to make sure that the data are fit for such analysis. 1) More precisely, you want to make sure that there are no missing values in the data. 2) Moreover, you want to ensure that each item's name in the "Description" is consistent. E.g., you want "Description" values such as " coffee black", "coffee &nbsp;black", " coffee black &nbsp;", etc. to be mapped to the same value (e.g. "coffee black"). 3) Last but not least, you want to remove transactions that were canceled. Such transactions correspond to rows where the invoice number starts with letter 'C'.

Apply these preprocessing steps to the dataset `retail_df` and apply them on the dataframe itself (e.g. set inplace=True).

In [ ]:
# your code

#### **a(iv)** 
After applying the preprocessing steps in **a(iii)** , repeat again the task **a(ii)**, that is:

- Show the number of rows in the dataset.
- Show the number of unique customers.
- Show the number of all unique product names.
- Show the list of all the countries where the customers reside.

Which values changed?

In [ ]:
# your code

**Your answer:** 

### b) Frequent itemsets and Association rules (8 pts)

#### **b(i)** 
Each invoice number in the dataset identifies a unique transaction. There are potentially many rows in the dataframe having the same invoice number. We want to analyze items that are frequently purchased together, that is, items that appear in the same transaction.

Create a new dataframe named `transaction_df` with two columns: "Invoice" and "Description". Here the "Invoice" value is the index of the dataframe (the unique number identifying each row) and "Description" is the column containing all items (without duplicates) purchased within the transaction with that invoice number. Display the first few rows of your dataframe. How many rows does the `transaction_df` have?

In [ ]:
# your code

**Your answer:**

#### **b(ii)** 
Next, we want to compute frequent itemsets and association rules based on the sets of items ordered together. Use the TransactionEncoder to transform `transaction_df` into a matrix such that the value in the i-th row and the j-th column is $True$ if the i-th itemset contains product j, and $False$ otherwise. Save the matrix into a dataframe named `transactions`. Display the shape of the matrix.

*Hint: Note that your dataframe 'transactions' must contain as many rows as there are invoice numbers and as many columns as there are unique products.*

In [ ]:
# your code

#### **b(iii)** 
Use the apriori method on `transactions` to obtain all frequent itemsets using min_support=0.01. Display all frequent itemsets that have at least three items. What support count does an itemset have for our case if it satisfies min_support=0.01?

In [ ]:
# your code

**Your answer:**

#### **b(iv)**
Now we will discover association rules from the frequent itemsets. Using only the frequent itemsets with min_support=0.01 (the ones obtained in **b(iii)**), generate different association rules using min_conf=0.6 and min_conf=0.9 as thresholds. Show the association rules for each of the thresholds. What do you notice w.r.t. the number of association rules produced for the different thresholds?

In [ ]:
# your code

**Your answer:**

#### **b(v)** 
From the association rules that satisfy the confidence threshold 0.6, select and show the two rules with the highest lift. What do you notice if you compare the two rules with each other?

In [ ]:
# your code

**Your answer:**

#### **b(vi)** 

In the analysis tasks in **b)**, an itemset consisted of items that had the same invoice number (same transaction items). Thus, if an itemset was frequent, it meant that the items in it were frequently purchased together.
An association rule $A \Rightarrow B$ meant that if items in $A$ are purchased, then the items in $B$ are also purchased in that same transaction.

Suppose that we would repeat the analysis, but this time, the itemsets would consist of items having the same "CustomerID" (bought from the same customer). Interpret the meaning of the frequent itemsets and association rules for this kind of itemsets.

**Your answer:**

### c) Sequence Mining (3 pts)

For this task, the dataset used is `retail_sequences.csv`. Run the cell below to save the dataset under the dataframe `retail_sequences`. Each row in the dataframe corresponds to a unique customer (from the retail dataset). The "Customer" column contains the customer ID, whereas the "Sequence" column contains the sequence of itemsets  purchased by that customer.  Each value of "Sequence" is a sequence (list) of itemsets $<I_1, I_2, ..., I_n>$. The items within the same itemset (list without duplicates) $I_i$ were purchased together (they had the same invoice number). The itemsets are ordered by the timestamp of the transaction (value of InvoiceDate).

In [ ]:
import pandas as pd 
retail_sequences = pd.read_csv('datasets/retail_sequences.csv', converters={'Sequence': pd.eval})

Given is the sequence *s= <{'lunch bag cars blue'}, {'herb marker rosemary','herb marker thyme'}, {'wooden star christmas scandinavian'}>*. Compute the support count of that sequence, that is, compute the number of customers whose corresponding itemset sequence contains it. Display its support count and the IDs of those customers.

*Hint: In the dataset provided, all product names are unified. They are all lowercase and have no trailing spaces.*

In [ ]:
s = [['lunch bag cars blue'], ['herb marker rosemary', 'herb marker thyme'] ,['wooden star christmas scandinavian']]

In [ ]:
# your code

## Question 4: Text Mining (12 pts)

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
from nltk.lm.preprocessing import padded_everygram_pipeline

### F.R.I.E.N.D.S.

In this task we will use the script from the well-known series "F.R.I.E.N.D.S.". We will apply feature extraction methods to map the line of each main character onto a vector of a vector space. Then we will train a classifier whose aim will be to predict the name of the character given a particular line from the script.
In the end, we will train language models using N-grams and produce fake sentences for each of the main characters.

### a) Data Loading and Preprocessing (4 pts)

#### **a(i)** 
Import the file `FRIENDS.csv` and save it into a dataframe named `friends_df`. Note that the dataframe must contain two columns: one indicating the character's name and one containing a line from the script. Display the first few lines from the dataframe.

<i>FYI: The script has been filtered so that it only contains lines from the main characters. The order of the lines in the data is the same as the order of the lines in the original script. Metadata and scene descriptions have been removed. Your corpus consists of all the lines contained in the data. Each row's "line" value is a single document. </i>

In [ ]:
# your code

#### **a(ii)**  
Plot the line count distribution among the six main characters (the six possible values of the column "character"). For example, show a plot containing one bar for each character whose height reflects the number of lines in `friends_df`. Briefly comment on the plot.

In [ ]:
# your code

**Your answer:**

#### **a(iii)**  
Create a corpus named `corpus` such that each document in the corpus corresponds to exactly one row's "line" in `friends_df`.

In [ ]:
# your code

#### **a(iv)**  
Write a function called `my_preprocessor` which, given a string, returns another string after tokenization, stopword removal and lemmatization have been applied. The remaining terms (tokens after stopword removal and lemmatization has been applied) should be joined in the same string using space ' '.

In [ ]:
# your code

#### **a(v)**   
You must apply your preprocessor `my_preprocessor` on each line contained in the `corpus`. Create a preprocessed corpus named `corpus_p` which contains the same lines as `corpus` after the preprocessor `my_preprocessor` has been applied.

In [ ]:
# your code

#### **a(vi)**   
Split the `friends_df` dataset from the previous task into training (80%) and test (20%) data preserving the distribution based on the "character" value. 

In [ ]:
# your code

#### **a(vii)**   
Similar to **a(iii)**, for the training data and the test data, create two corpora named `corpus_train` and `corpus_test` respectively. Each document in the training (test) corpus must correspond to exactly one row's "line" value in the corresponding training (test) dataframe.

In [ ]:
# your code

### b) Set of Words (4.5 pts)

#### **b(i)**  
We want to encode our text in such a way that for each word in the vocabulary, we are only interested in whether the word appears or not in a given document. Create such a Set of Words encoding for the whole corpus `corpus`. Use the previously defined preprocessor `my_preprocessor` as preprocessor.

In [ ]:
# your code

#### **b(ii)**  
Pick one (any) of the lines of the "line" column in the `friends_df` dataset. Display the line in:
    1) its original form, 
    2) its preprocessed version (the result contained in `corpus_p` after applying `my_preprocessor`), and 
    3) its encoding computed by the Set of Words method. This can be either an array (a vector) or a scipy matrix. 
Briefly comment on the Set of Words encoding.

In [ ]:
# your code

**Your answer:**

#### **b(iii)**  
Create a Set of Words encoding based only on the documents in `corpus_train`. Use the previously defined preprocessor `my_preprocessor` as a preprocessor.

In [ ]:
# your code

#### **b(iv)**  
In this task, we will use an SGD (Stochastic Gradient Descend) classifier to predict the character given a line from the corpus. Train the classifier on the Set of Words encoding of training corpus `corpus_train` using the character as the target feature and 'log_loss' as the loss function. Apply the classifier on the Set of Words encodings of both the training corpus and the test corpus `corpus_train`. Show its accuracy for both the training corpus and the test corpus.

In [ ]:
# your code

#### **b(v)**  
Briefly comment on the accuracy of the classifier compared to the expected accuracy of a random guesser (here: a model that simply guesses each character according to a distribution based on the line count). Use the line count distribution shown in **a(ii)** to reason about the approximate accuracy of the random guesser.

**Your answer:**

#### **b(vi)**  
Pick two lines from the dataset `friends_df`. Predict their character by applying the SGD classifier from **b(iv)** to their Set of Words encodings. Show the original lines, their original characters and the predicted characters.

In [ ]:
# your code

### c) Doc2Vec (1.5 pts)

#### **c(i)**  
In this part, we want to encode the lines using Doc2Vec. Create a Doc2Vec embedding based on the documents in the preprocessed corpus `corpus_p`. Set the vector dimension to 10 and min_count to 3.

In [ ]:
# your code

#### **c(ii)**  
Pick one (any) line from the dataset `friends_df`. Display the line and the character saying it. Find its most similar line w.r.t. the Doc2Vec encoding and display the original line and its corresponding character. Do the lines belong to the same character?

In [ ]:
# your code

**Your answer:**

### d) Language model using N-grams (2 pts)

#### For the following tasks, use the `friends_df` data.

#### **d(i)**  
For each character, create a corresponding corpus. Each corpus must be a list of documents. Each document corresponds to one "line" value of that character and it should be a list of terms. You must obtain this list of terms after applying preprocessing steps such as to lowercase, no punctuation, and tokenization to the original line. Do not perform stemming/lemmatization and/or stopword removal for this task.
Display the corpus of one of the characters.

In [ ]:
# your code

#### **d(ii)**  
For each character separately, build a trigram language model using MLE. Use both right and left padding and learn each language model using the character's corpus from **d(ii)**.

In [ ]:
# your code

#### **d(iv)**  
For each character, use the created trigram language model to generate a sentence of ten words. Display the sentences.

In [ ]:
# your code

## Question 5: Process Mining (22pts)

In [ ]:
import pandas as pd
import pm4py

In [ ]:
from pm4py.statistics.traces.generic.log import case_statistics

In [ ]:
from pm4py.algo.conformance.tokenreplay import algorithm as token_based_replay

In [ ]:
import matplotlib.pyplot as plt

### a) Loading the Data and Basic Statistics (9pts)

#### **a(i)** 
Load the data **reimburse.csv** and create a PM4Py event log. In doing so, use the following column mapping:
 - *Activity* is the activity key
 - *Case* is the case ID
 - *Timestamp* is the timestamp

In [ ]:
# your code

#### **a(ii)** 
Compute and print the following basic information:
- Number of events
- Number of cases
- Earliest timestamp
- Latest timestamp
- Number of trace variants

In [ ]:
# your code

#### **a(iii)** 
In process mining, multiple events that have the same timestamp can cause problems because the ordering of events (or even activities) becomes unclear. Moreover, they can indicate batching (i.e., one activity is executed for multiple cases simultaneously). Therefore, during your analysis, it is good to keep that in mind. To this end, compute the following statistics/answer the following questions:

1. How many events occur almost at the same time (i.e., within less than 100ms as the preceding event. (Proceeding event in the *entire* event log)?
2. Are there resources that complete two activities at the same time (within less than 100ms)?
3. How many cases are there in which two activities are executed at the same time (i.e., two events that belong to the same case occur within less than 100ms)?

*Hint: Depending on how you find the answers, be careful about event orderings.*

In [ ]:
# your code

#### **a(iv)** 
Provide a plot that shows the number of running cases (i.e., cases that have started but not yet finished) over time. You may assume that the log only contains complete traces. For each case that has started, its completion is the last observed event associated with that case. In case multiple cases start or end at the same time, you also generate multiple values.

In [ ]:
# your code

**Your answer**: *(Briefly describe the differences between the two models in about five sentences here.)*

### b) Discovery and Conformance Checking (9pts)

#### **b(i)** 
Mine a Petri net using Inductive Miner and display it.

In [ ]:
# your code

#### **b(ii)** 
Compute the fitness of the discovered Petri net using token-based replay.

*Hint: PM4Py can directly (using the top-level API) compute the number of missing, remaining, consumed, and produced tokens. Based on these, you can, for example, compute the token-based replay fitness.*

In [ ]:
# your code

#### **b(iii)** 
Filter the log to contain only traces where *Register Low* occurs. How many traces does the resulting log `log_low` contain?


In [ ]:
# your code

#### **b(iv)** 
Discover a Petri net for `log_low` and compute its fitness. How does this model differ from the model you discovered in *b(i)*? 

Suppose each of your produced process models is considered a 2-class classifier: provided a trace, it returns "Yes" if and only if the trace can be replayed by the model. Based on this perspective, how would the two process models compare in terms of precision?

In [ ]:
# your code

**Your answer:** *(Briefly describe the difference between the two models here. About two sentences can be enough.*)

**Your answer**: *(Relate your observations to precision here.)*

### Conformance Diagnostics De-jure Model
The process owner provides you a de-jure model (i.e., a model of the should-be process) and a slightly changed version of the so far considered event log. 

In this task, you will again apply conformance checking by means of token-based replay to provide diagnostics on deviations.

#### **b(v)** 
Load the Petri net *pn_conf.apnml*, the event log *log_conf.xes*, and provide the overall (i.e., model-based) token-based replay fitness score.



In [ ]:
# your code

#### **b(vi)** 
To provide additional diagnostics on the deviations, compute the missing, consumed, produced and remaining number of tokens for **each place**. 
To do so, use the following *pm4py* code:
    
    from pm4py.algo.conformance.tokenreplay import algorithm as token_based_replay
	tbr_results, place_fitness, transition_fitness, notexisting_activities_in_model =
    token_based_replay.apply(log_conf, net_conf, im_conf, fm_conf, parameters={"enable_pltr_fitness": True, "disable_variants": True})

After running this line for log `log_conf`, Petri net `net_conf` with initial marking `im_conf` and final marking `fm_conf`, the variable `place_fitness` will contain the token counts for each place and trace. Therefore, you will only need aggregate over the traces.
Print a table of the token counts per place. 


In [ ]:
# your code

#### **b(vii)** 
Consider the token counts per place and a few unfitting traces, which deviation(s) do you observe? Describe the deviation and briefly explain how it can be related to the token counts of the individual places. For example, activity *xy* is often missing resulting in a high number of missing tokens in place *p*.  


In [ ]:
# your code

**Your answer:** *(Describe the deviation(s). One sentence can already be enough.)*

**Your answer:** (*Relate the deviation(s) to the token counts of the individual places. Roughly five sentences can be enough for a precise description.*)

### c) Analyzing Fraud (4pts)

#### **c(i)** 
Filter the event log so that it only contains traces where a fraud report is filled (occurrence of `Fill Fraud Report`). For theses traces, create a bar plot showing the number of products of a certain brand involved in the fraud. Describe the resulting plot.

*Hint: Each case is associated with precisely one brand.*


In [ ]:
# your code

**Your answer:** (*Describe the plot in two to three sentences.*)

#### **c(ii)**  
The plot shows differences between brands. Discuss the result. Consider what you learned in Lecture 11 (association rules). Try to provide additional analysis results to underpin your discussion.

*Hint: A very short additional analysis (i.e., a few lines of code) might already be sufficient.*


In [ ]:
# your code (for a short additional analysis)

**Your answer:** *(Relate your results to Lecture 11, approximately one short paragraph)*

## Question 6 - Simpson's Paradox (3 pts)

### Sex Bias in Berkeley Graduate Admissions?

In the Fall of 1973, the University of California at Berkeley released data about their graduate class. The data showed the major the applicant applied to, their self-reported gender (Male or Female), and whether or not they were accepted or rejected. The acceptance rates between men and women were different. This caused immediate concern in the public as people thought that Berkeley was biased against women.

The "Berkeley Dataset" contains all 12,763 applicants to UC-Berkeley's graduate programs in Fall 1973. This dataset was published by UC-Berkeley researchers in an analysis to understand the possible gender bias in admissions.

Dataset Variables:

Year : number ➜ The application year (this value is always 1973)

Major : string ➜: An anonymized major code (either A, B, C, D, E, F, or Other). The specific majors are unknown except that A-F are the six majors with the most applicants in Fall 1973

Gender : string ➜ Applicant self-reported gender (either M or F)

Admission: string ➜ Admission decision (either Rejected or Accepted)

**a)**
Upload the data from the `berkeley.csv` file and load it into a dataframe named `data`. Display the first few lines from the dataframe.

In [ ]:
# your code

**b)** Remove the "Year" column as it does not contain any information in this dataset (all years are 1973.)

In [ ]:
# your code

**c)** For each of the values of column "Gender", compute the admission rate and compare them against each other.

In [ ]:
# your code

**Your answer:**

**d)** For each value combination of the "Gender" and "Major" columns, compute the admission rate. Compare the admission rate of women against the admission rate of men for each of the majors.

In [ ]:
# your code

**Your answer:**

**e)** Can you confirm there is a sex bias in the admission rates of the students?

**Your answer:**

## Question 7: Big Data (15pts)

In [ ]:
import pandas as pd

You are a data scientist at NASA, an agency for civil space programs, and with your team you develop and maintain the software of the NASA Crew Exploration Vehicle (CEV).  Your task is to analyze the performance of the software, and as a first exploratory step, you would like to **compute the mean execution times** of function calls within that software. Since the running vehicle will generate a high throughput of observable events in a stream, you decide to set up a MapReduce pipeline in Hadoop. 

The file **nasa-cev-software-tests.tsv** records timestamped events of the vehicle's software tests. The log contains the columns *Case*, *Activity* and *Timestamp*, denoting the case ID, the activity key (method call) and timestamp of the event record in nanoseconds, respectively. Furthermore, the log contains the columns *Lifecycle Transition* and *Execution ID*. The lifecycle transition takes either of the values *start* and *complete*, stating whether the corresponding activity (method call) in that row is being started or completed at the specified timestamp. The execution ID relates each event to a concrete method call, i.e., for each execution ID, there are exactly two entries (namely a *start* and a *complete* event) in the log.

### a) Plan the Maths (2pt):

The mean $\mu_n$ over numerical values $v_1,...,v_n$ is well-known to be computed as $\mu_n = \frac{1}{n}\sum_{i=1}^{n} v_i$.\
One may also use the alternative recursive formalization $\mu_{n+m} = \frac{n\cdot\mu_{n} + m\cdot\mu_{m}}{n+m}$. \
What is the advantage of using the alternative formalization when you think of handling streaming data or distributed data? Briefly explain.


**Your answer:**
1. In case of streaming data, the mean can be computed in real time by keeping the number of seen values.
2. In case of distributed data, the mean of all data can be computed from the mean of each nodes if the number of values of each node is known.

### b) Set up MapReduce (10pts)

Realize the computation of the mean execution times of activities as a MapReduce job. 
You need to implement this a cascaded MapReduced job. This means that the output of the first job will serve as the input of the second job. In the first job, derive the execution times of each activity execution, i.e. the time difference between the *complete* and the *start* lifecycle transition of each activity execution. In the second step, aggregate this timing information to compute the mean.

#### **b(i)** 
Specify the *function signatures* of the map functions and the reduce functions that you are going to use.\
I.e., find concrete sets to substitute $K_1, V_1, ... $ in the general signatures for map and reduce functions \

$ map:  K_1 \times V_1 \rightarrow (K_2 \times V_2)^* $\
$ reduce: K_2 \times (V_2)^* \rightarrow (V_3)^*$ (or a singleton $V_3$) 

*Hint: You may introduce symbols to denote sets, e.g. $Act$ for the set of activities.\
You may also first implement the solution (b(ii)) to get an idea about the underlying signatures.\
Mind that you need two map and two reduce functions.*

**Your answer:**

#### **b(ii)**: 
Specify map functions and reduce functions to compute the mean execution time per activity as python scripts.

In [ ]:
# your code (nasa_mapper1.py)

In [ ]:
# your code (nasa_reducer1.py)

In [ ]:
# your code (nasa_mapper2.py)

In [ ]:
# your code (nasa_reducer2.py)

### c) Run MapReduce (3pts)

In the following, please use one of your team members' matriculation number as an identifier.

#### **c(i) (Randomization)**: 
Before applying your functions from the previous step to the dataset, please insert the matriculation number and run the following lines to randomly filter out a few of the traces in the event log, and continue working with the filtered log. 


In [ ]:
# your matriculation number here
matr_nb = # ... #

# utility code (do not change)
import random
random.seed(matr_nb)

full_df = pd.read_csv("datasets/nasa-cev-software-tests.tsv", sep="\t")

case_ids = list(set(full_df["Case"].values))
case_ids.sort()
filtered_out_case_ids = random.sample(case_ids, 10)
filtered_case_ids = [case_id for case_id in case_ids if case_id not in filtered_out_case_ids]
randomized_df = full_df[full_df["Case"].isin(filtered_case_ids)]

randomized_df.to_csv("datasets/nasa-cev-software-tests-randomized-" + str(matr_nb) + ".tsv",
          columns=["Activity", "Timestamp", "Lifecycle Transition", "Execution ID"],
          sep="\t", index=False, header=False)

#### **c(ii)**:
Now, it is time to work with the Hadoop Distributed File System (HDFS). Follow the instructions below and show your results in each step (screenshots of the command line).

    1) Import the event log to your Docker engine (at /usr/local/hadoop/(your_matr_nb)-event-log/). You also need to import the python scripts, but only document the event log import here.
    2) Upload the files to the running HDFS (at /input/(your_matr_nb)-event-log/).
	3) Run Hadoop commands for the MapReduce computation.
    4) Show the final output (computed mean execution times).

In [ ]:
from IPython.display import Image
# your code
# Image(filename='filename_screenshot_of_a1')

In [ ]:
# your code
# Image(filename='filename_screenshot_of_a2')

In [ ]:
# your code
# Image(filename='filename_screenshot_of_a3')

In [ ]:
# your code
# Image(filename='filename_screenshot_of_a4')